src/flow/src/df_optimizer.rs

```diff
@@ -0,0 +1,599 @@
+        Arc::new(CheckGroupByRule::new()),
+        Arc::new(TypeCoercion::new()),
+    ]);
+    let plan = analyzer
+        .execute_and_check(plan, &cfg, |p, r| {
+            debug!("After apply rule {}, get plan: \n{:?}", r.name(), p);
+        })
+        .context(DatafusionSnafu {
+            context: "Fail to apply analyzer",
+        })?;
+
+    let ctx = OptimizerContext::new();
+    let optimizer = Optimizer::with_rules(vec![
+        Arc::new(OptimizeProjections::new()),
+        Arc::new(CommonSubexprEliminate::new()),
+        Arc::new(SimplifyExpressions::new()),
+        Arc::new(UnwrapCastInComparison::new()),
+    ]);
+    let plan = optimizer
+        .optimize(plan, &ctx, |_, _| {})
+        .context(DatafusionSnafu {
+            context: "Fail to apply optimizer",
+        })?;
+
+    Ok(plan)
+}
+
+/// To reuse existing code for parse sql, the sql is first parsed into a datafusion logical plan,
+/// then to a substrait plan, and finally to a flow plan.
+pub async fn sql_to_flow_plan(
+    ctx: &mut FlownodeContext,
+    engine: &Arc<dyn QueryEngine>,
+    sql: &str,
+) -> Result<TypedPlan, Error> {
+    let query_ctx = ctx.query_context.clone().ok_or_else(|| {
+        UnexpectedSnafu {
+            reason: "Query context is missing",
+        }
+        .build()
+    })?;
+    let stmt = QueryLanguageParser::parse_sql(sql, &query_ctx)
+        .map_err(BoxedError::new)
+        .context(ExternalSnafu)?;
+    let plan = engine
+        .planner()
+        .plan(stmt, query_ctx)
+        .await
+        .map_err(BoxedError::new)
+        .context(ExternalSnafu)?;
+    let LogicalPlan::DfPlan(plan) = plan;
+
+    let opted_plan = apply_df_optimizer(plan).await?;
+
+    // TODO(discord9): add df optimization
+    let sub_plan = DFLogicalSubstraitConvertor {}
+        .to_sub_plan(&opted_plan, DefaultSerializer)
+        .map_err(BoxedError::new)
+        .context(ExternalSnafu)?;
+
+    let flow_plan = TypedPlan::from_substrait_plan(ctx, &sub_plan).await?;
+
+    Ok(flow_plan)
+}
+
+struct AvgExpandRule {}
+
+impl AvgExpandRule {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl AnalyzerRule for AvgExpandRule {
+    fn analyze(
+        &self,
+        plan: datafusion_expr::LogicalPlan,
+        _config: &ConfigOptions,
+    ) -> datafusion_common::Result<datafusion_expr::LogicalPlan> {
+        let transformed = plan
+            .transform_up_with_subqueries(expand_avg_analyzer)?
+            .data
+            .transform_down_with_subqueries(put_aggr_to_proj_analyzer)?
+            .data;
+        Ok(transformed)
+    }
+
+    fn name(&self) -> &str {
+        "avg_expand"
+    }
+}
+
+/// lift aggr's composite aggr_expr to outer proj, and leave aggr only with simple direct aggr expr
+/// i.e.
+/// ```ignore
+/// proj: avg(x)
+/// -- aggr: [sum(x)/count(x) as avg(x)]
+/// ```
+/// becomes:
+/// ```ignore
+/// proj: sum(x)/count(x) as avg(x)
+/// -- aggr: [sum(x), count(x)]
+/// ```
```

Comment:
You can do some unitests to see if the results are as expected

Reply:
> You can do some unitests to see if the results are as expected

unit test for this is `test_tumble_composite` in `transform/aggr.rs`